integr8ly/grafana-operator
# INTLY-7282,5821,8371 Allow GrafanaDashboard CRs to be reconciled by more than 1 grafana-operator,Implement folder support, custom UID generation

## Jira 
Allow GrafanaDashboard CRs to be reconciled by more than 1 operator : https://issues.redhat.com/browse/INTLY-7282
Generate UID with md5sum for each dashboard : https://issues.redhat.com/browse/INTLY-5821
Implement folder support : https://issues.redhat.com/browse/INTLY-8371

## Description
To accommodate the move to cluster monitoring from the Application Monitoring Operator for Integreatly we need to ensure that dashboardCR's can be reconciled by more than 1 grafana operator and be found in any namespace, to enable this we need to ensure that the dashboards don't provide their own `UID` which we now create using a md5sum of the dashboard name and namespace.
To prevent dashboards encountering naming conflicts with dashboards from other namespaces they are now separated and grouped into their respective namespace named folder in each grafana instance,

## Verification
- Spin up an OSD cluster and login through your CLI.
- Follow the [manual procedure to install grafana](https://github.com/integr8ly/grafana-operator/blob/master/documentation/deploy_grafana.md#manual-procedure).
- [Deploy the Grafana CR](https://github.com/integr8ly/grafana-operator/blob/master/documentation/deploy_grafana.md#deploying-grafana).
- Create a few namespaces and dashboards in each `oc create -f deploy/examples/dashboards/SimpleDashboard.json -n <namespace>`.
- Navigate to the `Grafana` CR and note the uid of the dashboard present in the dashboard list .
- Ensure that the dashboards appear in the Grafana UI.
- Check out this branch.
- Change the ORG environment variable in the MakeFile to your repo name on quay.io.
- Run `make image/build` .
- Run `QUAY_USERNAME=<your user> QUAY_PASSWORD=<your password> REPO=<your quay name> make image/push`.
- Log into Quay and make your grafana-operator repository public.
- In OSD dashboard navigate to the `grafana-operator` deployment and substitute the the image link from integreatly to the one you just pushed.
- Navigate to the Grafana CR in the Grafana namespace, ensure that the dashboard list is reconciled with dashboards from other namespaces and that the previously noted uid has changed to a md5sum hash value.
- Navigate to the previously deployed `grafanadashboard` CRs and ensure that the status block has been removed.
- Navigate to the Grafana UI and ensure that dashboards appear in namespaced folders.
- Ensure that the UID in the URL of any dashboard is also in form of an md5sum hash value.
